### ISSUE TITLE
auction demo example with no submitting transactions fails

### ISSUE THREAD

**Issue by piachristel** [hyperledger-labs/fabric-private-chaincode#42]
If one creates an auction, closes it and evaluates it (so without having submitted any bits), an error is thrown and the peer stops.
[error.txt](https://github.com/hyperledger-labs/fabric-secure-chaincode/files/3105833/error.txt)

**Comment by mbrandenburger** [comment]
I cannot reproduce this error with the current code base. I suggest we close it and if needed reopen again.

**Comment by annumberhocker** [comment]
This is still a problem.  Scenario is to create and close an auction with no bids, the create a a new auction.  The peer crashes
```
 >>>> Evaluate auction. Response should be NO_BIDS 
/home/bcuser/work/src/github.com/hyperledger-labs/fabric-private-chaincode/integration/../fabric/bin//peer.sh chaincode invoke -o localhost:7050 -C mychannel -n auction_test -c {"Args":["[\"eval\",\"MyAuction3\"]", ""]}
2019-08-23 17:12:11.147 UTC [chaincodeCmd] chaincodeInvokeOrQuery -> INFO 001 Chaincode invoke successful. result: status:200 payload:"{\"ResponseData\":\"Tk9fQklEUw==\",\"Signature\":\"MEUCIAGXJSt+VUirxZ+6Ex+qFGDN4vY+hla3qjV+GmwcSJjYAiEAjcgOU4S5UlHbDZsLFTiTHmHZchNFRIR0hv5LCY0PDtw=\",\"PublicKey\":\"MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEXRDkJgANkhO2fiPA488fELfP00gqyTNVJb3ZX50pHQTFdR2pj1HNLxz8U+gF5/UmkFfpxavi4uPBjQKni1W/mA==\"}"
auction_test.sh: NO_BIDS
 PASSED 
 >>>> Create a new auction. Response should be OK 
/home/bcuser/work/src/github.com/hyperledger-labs/fabric-private-chaincode/integration/../fabric/bin//peer.sh chaincode invoke -o localhost:7050 -C mychannel -n auction_test -c {"Args": ["[\"create\",\"MyAuction4\"]", ""]} --waitForEvent
 auction_test.sh: test failed: /home/bcuser/work/src/github.com/hyperledger-labs/fabric-private-chaincode/integration/../fabric/bin//peer.sh chaincode invoke -o localhost:7050 -C mychannel -n auction_test -c {"Args": ["[\"create\",\"MyAuction4\"]", ""]} --waitForEvent 
/home/bcuser/work/src/github.com/hyperledger-labs/fabric-private-chaincode/integration/../fabric/bin//lib/common_ledger.sh: line 143: kill: (23979) - No such process
Makefile:14: recipe for target 'auction_test' failed
make: *** [auction_test] Error 111
```
Here is the output from the debugger: 
```
2019-08-23 17:12:11.387 UTC [endorser] callChaincode -> INFO 128 [mychannel][de55a866] Entry chaincode: name:"ecc_auction_test" 
2019-08-23 17:12:11.387 UTC [peer.chaincode.dev_test-jdoe_test-ecc_auction_test-0] func2 -> INFO 129 2019-08-23 17:12:11.387 UTC [ecc] Debugf -> DEBU 01a Invoke is running [["create","MyAuction4"]]
2019-08-23 17:12:11.389 UTC [endorser] callChaincode -> INFO 12a [mychannel][de55a866] Exit chaincode: name:"ecc_auction_test"  (3ms)
2019-08-23 17:12:11.390 UTC [comm.grpc.server] 1 -> INFO 12b unary call completed grpc.service=protos.Endorser grpc.method=ProcessProposal grpc.peer_address=127.0.0.1:45398 grpc.code=OK grpc.call_duration=3.79826ms
2019-08-23 17:12:13.154 UTC [gossip.privdata] StoreBlock -> INFO 12c [mychannel] Received block [14] from buffer
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x7f265eb560e1]

goroutine 1764 [running]:
plugin/unnamed-f95e718e88aabe57ce9c9f20938aaa4ee662b9d9.(*VSCCECC).checkEnclaveEndorsement(0xc000248640, 0xc000259350, 0xc002f31960, 0x0, 0x0)
	/home/bcuser/work/src/github.com/hyperledger-labs/fabric-private-chaincode/ecc/vscc/ecc_validation_logic.go:171 +0xa71
plugin/unnamed-f95e718e88aabe57ce9c9f20938aaa4ee662b9d9.(*VSCCECC).Validate(0xc000248640, 0xc002d9f600, 0xd54, 0xd80, 0xc002dbe920, 0x19, 0x20, 0x1, 0x1)
	/home/bcuser/work/src/github.com/hyperledger-labs/fabric-private-chaincode/ecc/vscc/ecc_validation_logic.go:101 +0x5c1
plugin/unnamed-f95e718e88aabe57ce9c9f20938aaa4ee662b9d9.(*ECCValidation).Validate(0xc000248620, 0xc00004d080, 0xc002d4f400, 0x10, 0x0, 0x0, 0xc00285b4a0, 0x1, 0x1, 0xc0000e3b30, ...)
	/home/bcuser/work/src/github.com/hyperledger-labs/fabric-private-chaincode/ecc/vscc/ecc_validation_plugin.go:70 +0x39a
github.com/hyperledger/fabric/core/committer/txvalidator.(*PluginValidator).ValidateWithPlugin(0xc003102f00, 0xc002e35e00, 0x3, 0x3)
	/home/bcuser/work/src/github.com/hyperledger/fabric/core/committer/txvalidator/plugin_validator.go:99 +0x299
github.com/hyperledger/fabric/core/committer/txvalidator.(*VsccValidatorImpl).VSCCValidateTxForCC(0xc00301f540, 0xc002e35e00, 0x1, 0x1)
	/home/bcuser/work/src/github.com/hyperledger/fabric/core/committer/txvalidator/vscc_validator.go:272 +0xc0
github.com/hyperledger/fabric/core/committer/txvalidator.(*VsccValidatorImpl).VSCCValidateTx(0xc00301f540, 0x0, 0xc0000d4d00, 0xc002d9f600, 0xd54, 0xd80, 0xc00004d080, 0x4, 0x49, 0xc002dc2c78)
	/home/bcuser/work/src/github.com/hyperledger/fabric/core/committer/txvalidator/vscc_validator.go:216 +0xeb6
github.com/hyperledger/fabric/core/committer/txvalidator.(*TxValidator).validateTx(0xc0030e7500, 0xc002dc2f70, 0xc002dc96e0)
	/home/bcuser/work/src/github.com/hyperledger/fabric/core/committer/txvalidator/validator.go:345 +0xf00
github.com/hyperledger/fabric/core/committer/txvalidator.(*TxValidator).Validate.func1.1(0xc0030e7500, 0xc00004d080, 0xc002dc96e0, 0x0, 0xc002d9f600, 0xd54, 0xd80)
	/home/bcuser/work/src/github.com/hyperledger/fabric/core/committer/txvalidator/validator.go:158 +0xcb
created by github.com/hyperledger/fabric/core/committer/txvalidator.(*TxValidator).Validate.func1
	/home/bcuser/work/src/github.com/hyperledger/fabric/core/committer/txvalidator/validator.go:155 +0x11b
bcuser@sys2-poc-bc-cf:~/work/src/github.com/hyperledger-labs/fabric-private-chaincode/integration$ 
```

**Comment by annumberhocker** [comment]
Per Michael: it fails in line 171 in going through range-queries which shouldn't happen during a create (i.e., line 170 should go immediately to 178).  Maybe we don't clean properly/reset our rwsets between invocations?

Per Marcus: maybe the problem is an empty result of a range query

**Comment by g2flyer** [comment]
it's kind of strange that a range-query would make create fail which should not involve range (composite) queries. But maybe it fails already with the previous tx (which if i understand correctly is an eval for an auction with no bid which would result in a composite query with empty result) and we just notice the effect of it due to asynchronous behavior of transaction processing only in the subsequent transaction?

also looking at processing of read-sets, there is certainly a bit strange what we do in shim.ccp: for normal get_state we record the query, regardless of result, whereas for composite_keys we report results (and potentially nothing if no result!) which at the very least looks inconsistent

**Comment by mbrandenburger** [comment]
PR #159 addresses this issue

**Comment by mbrandenburger** [comment]
@g2flyer 

I see your point regarding our creation of the readset in shim.cpp.

As you said, for normal reads we add the key to the readset. https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/ecc_enclave/enclave/shim.cpp#L44

But for range queries we only add a read if the range query result is not empty. https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/ecc_enclave/enclave/shim.cpp#L154

I would say this is not really clean but does it have a security impact? WDYT?

**Comment by g2flyer** [comment]
>
> I would say this is not really clean but does it have a security impact? WDYT?

It is is a security/functionality issue as it might prevent a conflict from being detected if in the same batch of transaction somebody else was modifying the state such that it actually would have returned something?

**Comment by mbrandenburger** [comment]
closed with #159
